ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java

```diff
@@ -0,0 +1,899 @@
+   * @param srcValue the {@link IndexValue} obtained from the application index.
+   * @return {@code true} if the record already exists in the target log, {@code false} otherwise.
+   * @throws StoreException if there is any problem with using the index.
+   */
+  private boolean alreadyExistsInTgt(StoreKey key, IndexValue srcValue) throws StoreException {
+    IndexValue tgtValue = tgtIndex.findKey(key);
+    return tgtValue != null && srcValue.isFlagSet(IndexValue.Flags.Delete_Index) == tgtValue.isFlagSet(
+        IndexValue.Flags.Delete_Index);
+  }
+
+  /**
+   * Copies the given {@code srcIndexEntries} from the given log segment into the swap spaces.
+   * @param logSegmentToCopy the {@link LogSegment} to copy from.
+   * @param srcIndexEntries the {@link IndexEntry}s to copy.
+   * @param lastModifiedTimeSecs the last modified time of the source index segment.
+   * @return @code true} if all the records  were copied. {@code false} if some records were not copied either because
+   * there was no more capacity or because a shutdown was initiated.
+   * @throws IOException if there were I/O errors during copying.
+   * @throws StoreException if there are any problems reading or writing to store components.
+   */
+  private boolean copyRecords(LogSegment logSegmentToCopy, List<IndexEntry> srcIndexEntries, long lastModifiedTimeSecs)
+      throws IOException, StoreException {
+    boolean copiedAll = true;
+    long totalCapacity = tgtLog.getCapacityInBytes();
+    FileChannel fileChannel = Utils.openChannel(logSegmentToCopy.getView().getFirst(), false);
+    long writtenLastTime = 0;
+    for (IndexEntry srcIndexEntry : srcIndexEntries) {
+      IndexValue srcValue = srcIndexEntry.getValue();
+      long usedCapacity = tgtIndex.getLogUsedCapacity();
+      if (isActive && (tgtLog.getCapacityInBytes() - usedCapacity >= srcValue.getSize())) {
+        fileChannel.position(srcValue.getOffset().getOffset());
+        Offset endOffsetOfLastMessage = tgtLog.getEndOffset();
+        // call into diskIOScheduler to make sure we can proceed (assuming it won't be 0).
+        diskIOScheduler.getSlice(LOG_SEGMENT_COPY_JOB_NAME, LOG_SEGMENT_COPY_JOB_NAME, writtenLastTime);
+        tgtLog.appendFrom(fileChannel, srcValue.getSize());
+        FileSpan fileSpan = tgtLog.getFileSpanForMessage(endOffsetOfLastMessage, srcValue.getSize());
+        IndexValue tgtValue;
+        if (srcValue.isFlagSet(IndexValue.Flags.Delete_Index)) {
+          IndexValue putValue = tgtIndex.findKey(srcIndexEntry.getKey());
+          if (putValue != null && putValue.getOffset().getName().equals(fileSpan.getStartOffset().getName())) {
+            tgtValue = new IndexValue(putValue.getSize(), putValue.getOffset(), putValue.getExpiresAtMs(),
+                srcValue.getOperationTimeInMs(), srcValue.getServiceId(), srcValue.getContainerId());
+            tgtValue.setNewOffset(fileSpan.getStartOffset());
+            tgtValue.setNewSize(srcValue.getSize());
```

Comment:
This code is the same as `PersistentIndex.markAsDeleted()`. `setNewOffset()` also has the logic to set the original message offset correctly - so its useful to do it this way.